patches/chromium/network_service_allow_remote_certificate_verification_logic.patch

```diff
@@ -0,0 +1,176 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Jeremy Apthorp <nornagon@nornagon.net>
+Date: Wed, 8 May 2019 17:25:55 -0700
+Subject: network service: allow remote certificate verification logic
+
+
+diff --git a/services/network/network_context.cc b/services/network/network_context.cc
+index 5d871fc08218b71c93141963db66465f775b1bc2..9e04684bede95aaa3a32c4b3aa280ff99d914a4c 100644
+--- a/services/network/network_context.cc
++++ b/services/network/network_context.cc
+@@ -310,6 +310,75 @@ std::string HashesToBase64String(const net::HashValueVector& hashes) {
+ 
+ }  // namespace
+ 
++class RemoteCertVerifier : public net::CertVerifier {
++ public:
++  RemoteCertVerifier(std::unique_ptr<net::CertVerifier> upstream): upstream_(std::move(upstream)) {
++  }
++  ~RemoteCertVerifier() override = default;
++
++  void Bind(mojom::CertVerifierClientPtr client_info) {
++    client_ = std::move(client_info);
++  }
++
++  // CertVerifier implementation
++  int Verify(const RequestParams& params,
++             net::CertVerifyResult* verify_result,
++             net::CompletionOnceCallback callback,
++             std::unique_ptr<Request>* out_req,
++             const net::NetLogWithSource& net_log) override {
++    out_req->reset();
++
++    net::CompletionOnceCallback callback2 = base::BindOnce(
++        &RemoteCertVerifier::OnRequestFinished, base::Unretained(this),
++        params, std::move(callback), verify_result);
++    int result = upstream_->Verify(params, verify_result,
++                                   std::move(callback2), out_req, net_log);
++    if (result != net::ERR_IO_PENDING) {
++      // Synchronous completion
```

Comment:
what happens with synchronous results ?

Reply:
synchronous results means the result was served from the cache. `upstream_` is a `CachingCertVerifier`, which will serve up a result from its cache synchronously and delete its callback if there's a cache entry present. It means your handler will only get called every so often rather than on every request, which I think is reasonable.

Reply:
Does the cache use result provided previously by the user or is it based on the results of the upstream verifier ?

Reply:
That's a great question. I think it'll use the result provided, but I'll write a test for it.

Reply:
bleh, it only caches the results of the upstream verifier. i'll have to ponder for a bit to figure out what to do here. it's a bit tricky because this code `std::move()`s the reference to the callback into the upstream `Verify` call, which means it's no longer accessible when in the sync branch. I could do some nonsense with `AdaptCallbackForRepeating` but it would be a bit of a hack...

Reply:
Since `RemoteCertVerifier` is the one finally calling into the response callback in `OnRemoteResponse`, we could store it as a class member rather than binding it to different functions. And in the synchronous result, we can directly call `OnRequestFinished`

Reply:
I don't think so; there's no guarantee that only one request is outstanding at a time.

Reply:
Ah forgot that the `CertVerifier` has a request queue, you are right in that case this won't work.

Reply:
hmm, maybe i can try putting the `RemoteCertVerifier` higher in the verifier chain than the caching verifier.